temporalio/client.py

```diff
@@ -1984,114 +2095,1664 @@ async def __anext__(self) -> temporalio.api.history.v1.HistoryEvent:
+                within.
+            rpc_timeout: Optional RPC deadline to set for the RPC call. This is
+                for each call made within, not overall.
+        """
+        await self._client._impl.update_schedule(
+            UpdateScheduleInput(
+                id=self.id,
+                updater=updater,
+                rpc_metadata=rpc_metadata,
+                rpc_timeout=rpc_timeout,
+            ),
+        )
 
 
 @dataclass
-class CancelWorkflowInput:
-    """Input for :py:meth:`OutboundInterceptor.cancel_workflow`."""
+class ScheduleSpec:
+    """Specification of the times scheduled actions may occur.
 
-    id: str
-    run_id: Optional[str]
-    first_execution_run_id: Optional[str]
-    rpc_metadata: Mapping[str, str]
-    rpc_timeout: Optional[timedelta]
+    The times are the union of :py:attr:`calendars`, :py:attr:`intervals`, and
+    :py:attr:`cron_expressions` excluding anything in :py:attr:`skip`.
 
+    .. warning::
+        Schedules are an experimental feature.
+    """
 
-@dataclass
-class DescribeWorkflowInput:
-    """Input for :py:meth:`OutboundInterceptor.describe_workflow`."""
+    calendars: Sequence[ScheduleCalendarSpec] = dataclasses.field(default_factory=list)
+    """Calendar-based specification of times."""
 
-    id: str
-    run_id: Optional[str]
-    rpc_metadata: Mapping[str, str]
+    intervals: Sequence[ScheduleIntervalSpec] = dataclasses.field(default_factory=list)
+    """Interval-based specification of times."""
+
+    cron_expressions: Sequence[str] = dataclasses.field(default_factory=list)
+    """Cron-based specification of times.
+    
+    This is provided for easy migration from legacy string-based cron
+    scheduling. New uses should use :py:attr:`calendars` instead. These
+    expressions will be translated to calendar-based specifications on the
+    server.
+    """
+
+    skip: Sequence[ScheduleCalendarSpec] = dataclasses.field(default_factory=list)
+    """Set of matching calendar times that will be skipped."""
+
+    start_at: Optional[datetime] = None
+    """Time before which any matching times will be skipped."""
+
+    end_at: Optional[datetime] = None
+    """Time after which any matching times will be skipped."""
+
+    jitter: Optional[timedelta] = None
+    """Jitter to apply each action.
+
+    An action's scheduled time will be incremented by a random value between 0
+    and this value if present (but not past the next schedule).
+    """
+
+    time_zone_name: Optional[str] = None
+    """IANA time zone name, for example ``US/Central``."""
+
+    @staticmethod
+    def _from_proto(spec: temporalio.api.schedule.v1.ScheduleSpec) -> ScheduleSpec:
+        return ScheduleSpec(
+            calendars=[
+                ScheduleCalendarSpec._from_proto(c) for c in spec.structured_calendar
+            ],
+            intervals=[ScheduleIntervalSpec._from_proto(i) for i in spec.interval],
+            cron_expressions=spec.cron_string,
```

Comment:
You should never receive a `cron_string` from the server. I'd consider removing that line.

Reply:
This will then always be an empty sequence of strings which is safe and still reads nice from a symmetry perspective. So which fields are populated remains a server decision.